File: Turquoise/UI/VC/TQArticleComposerViewController.m

#import "TQArticleComposerViewController.h"

#import "NSString+TQEncoding.h"
#import "TQNNTPManager.h"

@implementation TQArticleComposerViewController {
  TQNNTPManager *_nntpManager;
}

- (void)viewDidLoad {
  [super viewDidLoad];

  _nntpManager = [TQNNTPManager sharedInstance];

  UIColor *placeholderTextColor = [UIColor colorWithWhite:.6 alpha:1];
  _articleSubjectField.attributedPlaceholder =
      [[NSAttributedString alloc] initWithString:@"Subject:"
                                      attributes:@{ NSForegroundColorAttributeName : placeholderTextColor }];

  NSString *subject = @"";
  if (_parentArticle) {
    subject = _parentArticle.decodedSubject;
    if (![subject hasPrefix:@"Re: "]) {
      subject = [NSString stringWithFormat:@"Re: %@", subject];
    }
  }

  _articleSubjectField.text = subject;
}

- (void)dismissArticleComposer {
  [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)dismiss:(id)sender {
  [self dismissArticleComposer];
}

- (IBAction)postArticle:(id)sender {
  if ([_articleSubjectField.text tq_isEmpty]) {
    // TODO: alert
    NSLog(@"Cannot post message with empty subject");
    return;
  } else if ([_articleBodyView.text tq_isEmpty]) {
    // TODO: alert
    NSLog(@"Cannot post message with empty body");
    return;
  }

  NSString *subject = [_articleSubjectField.text tq_whitespaceAndNewlineStrippedString];
  NSString *message = [_articleBodyView.text tq_whitespaceAndNewlineStrippedString];
  TQNNTPArticle *article = [[TQNNTPArticle alloc] initWithSubject:subject
                                                          message:message
                                                        newsGroup:_newsGroup
                                                    parentArticle:_parentArticle];
    if (!article) {
      // something went wrong. cannot post article.
      // TODO: show error.
      NSLog(@"Not implemented yet!");
      return;
    }

  [_nntpManager postArticle:article completion:^(TQNNTPResponse *response, NSError *error) {
    if ([response isOk]) {
      NSLog(@"Article posted");
    } else {
      NSLog(@"Could not post article: %@", error);
    }

    [self dismissArticleComposer];
  }];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Turquoise/UI/VC/TQArticleComposerViewController.swift
import Foundation
import UIKit

class TQArticleComposerViewController : UIViewController {
  @IBOutlet var articleSubjectField: UITextField!
  @IBOutlet var articleBodyView: UITextView!

  var newsGroup: TQNNTPGroup?
  var parentArticle: TQNNTPArticle?

  override func viewDidLoad() {
    super.viewDidLoad()

    let placeholderTextColor = UIColor(white: 0.6, alpha: 1)
    self.articleSubjectField.attributedPlaceholder =
      NSAttributedString(string: "Subject:",
                         attributes: [ NSForegroundColorAttributeName : placeholderTextColor ])
    var subject = ""
    if let parentArticle = self.parentArticle {
      subject = parentArticle.decodedSubject
      if !subject.hasPrefix("Re: ") {
        subject = "Re: \(subject)"
      }
    }

    self.articleSubjectField.text = subject
  }

  func dismissArticleComposer() {
    self.dismiss(animated: true, completion: nil)
  }

  @IBAction func dismiss(_ sender: Any?) {
    self.dismissArticleComposer()
  }

  @IBAction func postArticle(_ sender: Any?) {
    guard let articleSubjectText = self.articleSubjectField.text, !articleSubjectText.tq_isEmpty else {
      // TODO: alert
      printError("Cannot post message with empty subject")
      return
    }
    guard let articleBodyText = self.articleBodyView.text, !articleBodyText.tq_isEmpty else {
      // TODO: alert
      printError("Cannot post message with empty body")
      return
    }

    let subject = articleSubjectText.tq_whitespaceAndNewlineStrippedString
    let message = articleBodyText.tq_whitespaceAndNewlineStrippedString
    guard let article = TQNNTPArticle(subject: subject,
                                message: message,
                                newsGroup: self.newsGroup,
                                parentArticle: self.parentArticle) else {
                                  // something went wrong. cannot post article.
                                  // TODO: show error.
                                  printError("Not implemented yet!")
                                  return
    }

    TQNNTPManager.sharedInstance.post(article: article) { (response, error) in
      if let response = response, response.isOk() {
        printInfo("Article posted")
      } else {
        printError("Could not post article: \(error != nil ? error.debugDescription : "")")
      }

      self.dismissArticleComposer()
    }
  }
}
